src/V3SplitVar.cpp

```diff
@@ -0,0 +1,1258 @@
+// DESCRIPTION: Verilator: Break variables into separate words to avoid UNOPTFLAT
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2020 by Wilson Snyder.  This program is free software; you can
+// redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+//
+// Verilator is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+//*************************************************************************
+// V3SplitVar divides a variable into multiple variables to avoid UNOPTFLAT warning
+// and get better perfomance.
+// Variables to be split must be marked by /*verilator split_var*/ metacomment.
+// There are sveral kinds of data types that may cause the warning.
+// 1) Unpacked arrays
+// 2) Packed arrays
+// 3) Unpacked structs
+// 4) Packed structs
+// 5) Bitfields within a signal. (Especially Verilog code predating structs/2D arrays.)
+// 2-5 above are treated as bitfields in verilator.
+//
+// What this pass does looks as below.
+//
+//     logic [1:0] unpcked_array_var[0:1] /*verilator split_var*/;
+//     always_comb begin
+//        unpacked_array_var[1] = unpacked_array_var[0]; // UNOPTFLAT warning
+//     end
+//     logic [3:0] packed_var /*verilator split_var*/;
+//     always_comb begin
+//        if (some_cond) begin
+//            packed_var = 4'b0;
+//        end else begin
+//            packed_var[3]   = some_input0;
+//            packed_var[2:0] = some_input1;
+//        end
+//     end
+//
+// is converted to
+//
+//     logic [1:0] unpcked_array_var0;
+//     logic [1:0] unpcked_array_var1;
+//     always_comb begin
+//        unpacked_array_var1 = unpacked_array_var0;
+//     end
+//     logic       packed_var__BRA__3__KET__;
+//     logic [2:0] packed_var__BRA__2_0__KET__;
+//     always_comb begin
+//        if (some_cond) begin
+//            {packed_var__BRA__3__KET__, packed_var__BRA__2_0__KET__} = 4'b0;
+//        end else begin
+//            packed_var__BRA__3__KET__   = some_input0;
+//            packed_var__BRA__2_0__KET__ = some_input1;
+//        end
+//     end
+// </pre>
+//
+//
+//*************************************************************************
+
+#include "config_build.h"
+#include "verilatedos.h"
+
+#include "V3Ast.h"
+#include "V3Const.h"
+#include "V3Global.h"
+#include "V3SplitVar.h"
+#include "V3Stats.h"
+
+#include <algorithm>  // sort
+#include <set>
+#include <vector>
+#include VL_INCLUDE_UNORDERED_MAP
+#include VL_INCLUDE_UNORDERED_SET
+
+static AstNodeAssign* newAssign(FileLine* fileline, AstNode* lhs, AstNode* rhs,
+                                const AstVar* varp) {
```

Comment:
Here and below please add e.g. "split" prefix to all global names to disambiguate, as sometimes all .cpp files are compiled in a single compiler run and it keeps things cleaner.

I'd however prefer either:

A. move these into a class just used to wrap the implementation internals (SplitVarImp).
B. Or, make a base class and inherit it into each Visitor.